climada/hazard/base.py

```diff
@@ -369,6 +371,361 @@ def set_vector(self, *args, **kwargs):
+        >>>         freq=(["time"], [0.4]),
+        >>>         event_id=(["time"], [4]),
+        >>>     ),
+        >>>     dict(
+        >>>         time=[datetime.datetime(2000, 1, 1)],
+        >>>         latitude=[0, 1],
+        >>>         longitude=[0, 1, 2],
+        >>>     ),
+        >>> )
+        >>> hazard = Hazard.from_raster_xarray(
+        >>>     dset,
+        >>>     data_vars=dict(
+        >>>         # Load frequency from 'freq' array
+        >>>         frequency="freq",
+        >>>         # Ignore 'event_id' array and use default instead
+        >>>         event_id="",
+        >>>         # 'fraction' array is loaded because it has the default name
+        >>>     ),
+        >>> )
+        >>> np.array_equal(hazard.frequency, [0.4]) and np.array_equal(
+        >>>     hazard.event_id, [1]
+        >>> )
+        True
+
+        """
+        # If the data is a string, open the respective file
+        if not isinstance(data, xr.Dataset):
+            LOGGER.info("Loading Hazard from file: %s", data)
+            data = xr.open_dataset(data)
+        else:
+            LOGGER.info("Loading Hazard from xarray Dataset")
+
+        # Initialize Hazard object
+        hazard = cls(haz_type=hazard_type)
+        hazard.unit = intensity_unit
+
+        # Update coordinate identifiers
+        coords = {"time": "time", "longitude": "longitude", "latitude": "latitude"}
+        if coordinate_vars is not None:
+            unknown_coords = [co for co in coordinate_vars if co not in coords]
+            if unknown_coords:
+                raise ValueError(
+                    f"Unknown coordinates passed: '{unknown_coords}'. Supported "
+                    "coordinates are 'time', 'longitude', 'latitude'."
+                )
+            coords.update(coordinate_vars)
+
+        # Retrieve dimensions of coordinates
+        dims = dict(
+            time=data[coords["time"]].dims,
+            longitude=data[coords["longitude"]].dims,
+            latitude=data[coords["latitude"]].dims,
+        )
+
+        # Stack (vectorize) the entire dataset into 2D (time, lat/lon)
+        # NOTE: We want the set union of the dimensions, but Python 'set' does not
+        #       preserve order. However, we want longitude to run faster than latitude.
+        #       So we use 'dict' without values, as 'dict' preserves insertion order
+        #       (dict keys behave like a set).
+        data = data.stack(
+            event=dims["time"],
+            lat_lon=dict.fromkeys(dims["latitude"] + dims["longitude"]),
+        )
+
+        # Transform coordinates into centroids
+        hazard.centroids = Centroids.from_lat_lon(
+            data[coords["latitude"]].values, data[coords["longitude"]].values
+        )
+
+        def to_csr_matrix(array):
+            """Store an array as sparse matrix, optimizing storage space"""
+            mat = sparse.csr_matrix(array)
+            mat.eliminate_zeros()
+            return mat
```

Comment:
I think with this method you cannot actually load sparse data (i.e. vector data). This is an issue imho.

Reply:
In my understanding, xarray does not consider sparse data. The arrays returned by xarray are always dense. So I don't see how this can be an issue here. I'm calling `eliminate_zeros` to make sure the resulting matrix is as sparse as possible.

Apparently, there you can use sparse data types as chunks, but that needs to be done on the user side and is not something xarray handles for you automatically, see e.g. https://stackoverflow.com/a/60451126

Reply:
Is it possible to store vector data in an xarray? So, basically saving/reading only the non-zero data points?

Reply:
Maybe I'm not sure what "vector data" actually means 🤔 You could store an array with only two dimensions "time" and "space" (basically the structure of `Hazard.frequency`), and then have two coordinates "latitude" and "longitude" that map each point in "space" to them. `from_raster_xarray` would read that as well.

To know if any data points are non-zero (or non-`NaN`), one has to load them, right? Xarray either loads all data at once or uses dask to devide the data into chunks that are only loaded into memory when needed.

Reply:
>Maybe I'm not sure what "vector data" actually means 🤔 You could store an array with only two dimensions "time" and "space" (basically the structure of Hazard.frequency), and then have two coordinates "latitude" and "longitude" that map each point in "space" to them. from_raster_xarray would read that as well.

Exactly, this is what I mean. But you are right, it first would have to be read... All right, let's see how we can solve this in the future.

Reply:
I at least found out the following: Xarray typically uses `NaN` to represent missing values, but `csr_matrix` stores `NaN`s explicitly. So I think it makes a lot of sense to replace `NaN`s with zeros before constructing the sparse matrix.

Reply:
Good idea! Although then I would suggest to use a more intelligent way to build the matrix. First, identify all the non-zero elements, and then use only these to construct the matrix. It is quite inefficient to first create a sparse row matrix and then alter its sparcity. Alternatively, first create a `coo_matrix`, change the sparcity, and then transform to a `csr_matrix`.

Reply:
I think this is not necessary. I just noticed that when you construct a matrix from some dense array, the zeros will be already eliminated (so we can save this extra call I put in). Plus, it seems like the constructor of a `csr_matrix` actually constructs a `coo_matrix` internally when being initialized from a dense array, see https://github.com/scipy/scipy/blob/v1.8.1/scipy/sparse/_compressed.py#L85

Reply:
All right! Thanks for checking. Let's just keep in mind that if we manage to actually deal with sparse data, it would be much more efficient to initialize the sparse matrix not from a dense array.

Reply:
NaN handling fixed and veryfied in abc9d320a2e970f2ee72cbd9e5bda1ecbe6f01e9